starcheck/src/lib/Ska/Starcheck/Obsid.pm

```diff
@@ -1998,9 +1997,13 @@ sub print_report {
 
             my $acq_prob = "";
             if ($c->{"TYPE$i"} =~ /BOT|ACQ/){
-                my $prob = $self->{acq_probs}->{$c->{"IMNUM${i}"}};
-                $acq_prob = sprintf("Prob Acq Success %5.3f", $prob);
-
+                if (defined $self->{acq_probs}->{$c->{"IMNUM${i}"}}){
```

Comment:
Why this new test for existence?  Don't we always have an acq prob for any acq star?  Anyway couldn't you do something like `my $prob = $self->{acq_probs}->{$c->{"IMNUM${i}"}} or 0.0`?  (If that works Perl).

Reply:
I think your "or" version is good.  I keep trying to anticipate "unitialized" errors but I guess you don't get one on the hash lookup in the "or" version.

With regard to why this new test for existence, I didn't think a comment would help here (when the information about the *change* wouldn't be useful to me in the future) and stuck it in the commit message ("Handle missing data (ReOpen schedule beginning) for proseco probs").  We have a bunch of missing probabilities for stars at the beginning of a replan/reopen (old probs were calculated on those stars, new probs are not).   I could just not print the observations but...

Reply:
I can't emphasize enough that code comments about why a particular test is done are *extremely valuable* to me and to your future self.  Basically during development you had some non-obvious insight into a particular situation that needs handling.  So explain that in detail *in the code* at the moment you think of it.  When we look at that line in 5 years we won't have to wonder.

You should not consider commit messages as a form of legacy documentation.  They provide some benefit during the code review process (though honestly I don't pay too much attention) but not so much later on.  Git blame does work and in some cases helps, but I would rather just see comments right in the code.

Reply:
I just haven't put as much time into trying to figure out the Perl.  I'm playing whack-a-mole with uninitialized warnings, and sometimes I even go crazy and change two things at once. If the warning goes away I stop playing.  That's why I thought comments wouldn't be much use to my future self.  

In this case, I thought your idea for the "or" worked on the undefined quantity, but now I'm seeing   "Use of uninitialized value $prob in sprintf at starcheck/src/lib/Ska/Starcheck/Obsid.pm line 2009." again so I'm going to go back to checking to see if the specific image probs is defined.  Probably don't have to set to to 0 for the hover-over though, can leave it as the empty string.

Reply:
I can't see any reason to waste time trying to make the code work for the replan/reopen section.  I have zero confidence in any of the starcheck results produced there and I never look at those, so you should just cut it.

Of course that would be hard in the context of these changes right now, so take that as something to do later and then clean up this code.

Reply:
The thing about comments: the more you struggle with a section of code and the more moles you whacked, the more important it is that you document that struggle because later on someone (e.g. me) can come along with a brilliant idea to tidy up the code and have no idea of the back-story of how it got there and what will break if I make some little change that seems reasonable.

Reply:
Confused by the comment, as I didn't even try to make the proseco probabilities get calculated for the replan/reopen observations (I started with that but stopped when you said so). I just made starcheck not spit out any *new* "uninitialized" warnings for those observations.  Since they were set up differently, that required some more handling and coding.  If we want to not print out *anything* in that section of the schedule, that's a different question.

Reply:
Also, my confusion was for your comment one up (in case there was still some confusion about replan reopen and proseco probabilities).

 Regarding comments in general, I think in most of this code it will break if you make some little change that seems reasonable; but I'm not arguing about adding comments, just stating that there's more work to doing the investigation to make a truly useful comment and I have no idea how much time to spend on any of the Perl.  I thought I was being somewhat judicious in the use of my time, given present time challenges, but will try to find a more verbose compromise.

Reply:
OK, and yeah, I'd just heard not calculating probabilities for the replan/reopen observations and ignoring their in-the-starcheck-output warnings, I hadn't heard starting the starcheck output at the beginning of the DOT.  That would be more a more complicated change.